Test interactions with sproxyd and integrate SPRXCLT-12

fmonjalet opened the pull request:
This PR sets up and uses an `sproxyd` container for cloudserver's multiple backend tests. This allows basic cloudserver / sproxyd integration testing at this level, and includes a non regression test for `SPRXCLT-12`.

It also reenables a number of disabled tests.

bert-e commented:
# Hello fmonjalet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-423 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.27`

* `8.8.1`


Please check the `Fix Version/s` of CLDSRV-423, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-423 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.29`

* `8.8.1`


Please check the `Fix Version/s` of CLDSRV-423, or the target
branch of this pull request.

Commit:
CLDSRV-423: refactor multiple backend put tests to avoid duplication

Commit:
CLDSRV-423: Add missing mock logger method

Commit:
CLDSRV-423: disable failing tests that have just been reenabled

They had been disabled for a long while and cannot be reenabled yet
because they don't pass, so keep on skipping them for now.

Tickets have been created to take care of them:

- CLDSRV-440
- CLDSRV-441
- CLDSRV-442
- CLDSRV-443

Commit:
CLDSRV-423: deploy sproxyd for multiple backend tests

Commit:
CLDSRV-423: add put test for sproxyd

Commit:
CLDSRV-423: reenable a good chunk of multiple backend tests

Commit:
CLDSRV-423: don't run sproxyd test when testing Ceph

Commit:
CLDSRV-423: perform two successive put in multiple backend tests

This tests for SPRXCLT-12 issue

Commit:
CLDSRV-423: reproduce SPRXCLT-12 more often

Commit:
CLDSRV-423: bump arsenal and sproxydclient to fix SPRXCLT-12

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-423 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.1`


Please check the `Fix Version/s` of CLDSRV-423, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-423 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.1`


Please check the `Fix Version/s` of CLDSRV-423, or the target
branch of this pull request.

francoisferrand commented on .github/docker/sproxyd/conf/fastcgi_params:
shoud we add all these config file to a "sproxyd test" image, so that is easy as well to deploy locally?

could be just a simple Dockerfile:

```
FROM sproxyd
ADD factcgi_params
ADD nginx.conf
RUN chown root:root ...
```

and we can add it as well as `build:` line in the docker-compose (in addition to the `image:`), to allow autoamtically building when needed, even if not pushed to the registry

fmonjalet replied:
Good idea, I did something in this spirit, tell me if it is what you expected.

francoisferrand commented on tests/multipleBackend/objectPut.js:
```diff
@@ -71,74 +74,99 @@ function put(bucketLoc, objLoc, requestHost, cb, errorDescription) {
+    putCases.forEach(testCase => {
+        it(`should put an object to ${testCase.name}`, done => {
+            async.series([
+                next => put(testCase.bucketLoc, testCase.objLoc, 'localhost', () => {
+                    checkPut(testCase);
+                    next();
+                }),
```
nit:
```suggestion
                next => put(testCase.bucketLoc, testCase.objLoc, 'localhost', next),
                next => {
                    checkPut(testCase);
                    next();
                }),
```

(and we may then merge it with the next `setTimeout` to simplify further)

jonathan-gramain commented on tests/multipleBackend/objectPut.js:
```diff
@@ -75,78 +75,78 @@ function put(bucketLoc, objLoc, requestHost, cb, errorDescription) {
-        });
-    });
-
-    it('should put an object to Azure based on object location', done => {
```
Not sure if this was wanted but this test doesn't seem to be carried over

fmonjalet replied:
I believe I just moved it somewhere else for consistency, it should be the `'azure'` entry of `putCases`. Tell me if I missed anything.

jonathan-gramain commented on tests/multipleBackend/objectPut.js:
```diff
@@ -157,6 +157,10 @@ describeSkipIfE2E('objectPutAPI with multiple backends', function testSuite() {
+                // Increase the probability of the first request having released
+                // the socket, so that it can be reused for the next request.
+                // This tests how HTTP connection reuse behaves.
+                next => setTimeout(next, 10),
```
I am a bit surprised that 10ms is enough for the socket to be closed, but maybe :)

fmonjalet replied:
It's not about closing the socket, but having the first request not use the socket anymore, so that the second can use it. It's pure js logic, not OS. This very small timeout seems to leave the time for the code responsible for this to be scheduled and release the socket, therefore making socket reuse much more frequent (checked by looking at nginx logs). Otherwise, the second request often uses a new socket instead of the existing one, defeating the purpose of the test.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
CLDSRV-423: provide a proper dockerfile for test sproxyd

Commit:
CLDSRV-423: improve async series usage in test

Commit:
CLDSRV-423: test distinct and overwriting PUTs

Commit:
Merge remote-tracking branch 'origin/development/8.8' into bugfix/CLDSRV-423-test-sproxyd

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-423 contains:


* `8.8.1`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.3`


Please check the `Fix Version/s` of CLDSRV-423, or the target
branch of this pull request.

JBWatenbergScality commented:
/status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

fmonjalet commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.8`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/8.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/8.7`



Please check the status of the associated issue CLDSRV-423.

Goodbye fmonjalet.
